### PR TITLE
Merge `conda-forge/main` into `rapidsai/main` (pt. 8)

### PR DESCRIPTION
**Pull request by jakirkham**
* Release 2.0.3 ( https://github.com/conda-forge/xgboost-feedstock/pull/132 )
* Use `noarch: python` for Python packages ( https://github.com/conda-forge/xgboost-feedstock/pull/144 )

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -13,11 +13,10 @@ source:
...
   patches:
     # xgboost patches
-    - 0001-conda-Unbundle-libxgboost.-dll-dylib-so.patch
     - 0001-Force-endian-flag-in-cross-compilation-mode.patch  # [arm64 or aarch64 or ppc64le]
-    - 0002-Use-static-cudart.patch
-    # xref: https://github.com/dmlc/xgboost/pull/9602
-    - 0003-numpy-ndarray-check.patch
+    - 0002-Enable-latest-libcxx-on-MacOS.patch  # [osx]
+    - 0003-Use-mingw-w64-path.patch
+    - 0004-Undo-dmlc-xgboost-9436.patch
```
Comment (jakirkham) [inline recipe/meta.yaml:19]: Could someone please double check these are all the right patches for us? Also please feel free to add/remove to bring these inline with expectations

File: recipe/build.sh [hunk 2]
```diff
@@ -4,9 +4,17 @@ set -exuo pipefail
...
+    XGB_CMAKE_ARGS=(-DOpenMP_C_FLAGS="-Xpreprocessor -fopenmp -I${PREFIX}/include" -DOpenMP_CXX_FLAGS="-Xpreprocessor -fopenmp -I${PREFIX}/include" -DOpenMP_C_LIB_NAMES=libomp -DOpenMP_CXX_LIB_NAMES=libomp -DOpenMP_libomp_LIBRARY=${PREFIX}/lib/libomp.dylib ${XGB_CMAKE_ARGS[@]+"${XGB_CMAKE_ARGS[@]}"} )
+fi
+
+
 if [[ ${cuda_compiler_version} != "None" ]]; then
-    export CMAKE_ARGS="-DUSE_CUDA=ON -DUSE_NCCL=ON -DBUILD_WITH_SHARED_NCCL=ON ${CMAKE_ARGS}"
-    export CMAKE_ARGS="-DBUILD_WITH_CUDA_CUB=ON -DPLUGIN_RMM=ON ${CMAKE_ARGS}"
+    XGB_CMAKE_ARGS=(-DUSE_CUDA=ON -DUSE_NCCL=ON -DBUILD_WITH_SHARED_NCCL=ON ${XGB_CMAKE_ARGS[@]+"${XGB_CMAKE_ARGS[@]}"} )
+    XGB_CMAKE_ARGS=(-DBUILD_WITH_CUDA_CUB=ON -DPLUGIN_RMM=ON ${XGB_CMAKE_ARGS[@]+"${XGB_CMAKE_ARGS[@]}"} )
 fi
```
Comment (jakirkham) [inline recipe/build.sh:18]: Could someone please verify we are passing the right flags here? Also please feel free to update if not

File: recipe/meta.yaml (hunk 1 above)
Comment (jakirkham) [inline recipe/meta.yaml:19]: Could someone please double check these are all the right patches for us?

Also please feel free to add/remove to bring these inline with expectations

Comment (hcho3) [reply recipe/meta.yaml:19]: Yes, they are all relevant to us.

**Review by hcho3 — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks Philip! 🙏

Is there something we need to pass to XGBoost to ensure `cudart` is statically linked?

Asking as [the CI log shows]( https://github.com/rapidsai/xgboost-feedstock/actions/runs/7581954367/job/20670338304?pr=46#step:4:1352 ):

```
2024-01-19T10:07:58.9852945Z WARNING (libxgboost,bin/xgboost): $RPATH/libcudart.so.12 not found in packages, sysroot(s) nor the missing_dso_whitelist.
```

**Comment by jakirkham** [comment]
Just a note on the missing DSO warning above, have converted all DSO warnings to errors with PR ( https://github.com/conda-forge/xgboost-feedstock/pull/154 ), which was merged in here with PR ( https://github.com/rapidsai/xgboost-feedstock/pull/47 )

As a result the same missing DSO warning we saw above, will now raise an error on CI

Tested this by opening a PR with the removed conda-forge patch ( https://github.com/rapidsai/xgboost-feedstock/pull/48 ). As expected this raises an error [on CI]( https://github.com/rapidsai/xgboost-feedstock/actions/runs/7591912499/job/20680595181?pr=48#step:4:1400 ):

```
    ['  ERROR (libxgboost,bin/xgboost): $RPATH/libcudart.so.12 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?', '  ERROR (libxgboost,lib/libxgboost.so): $RPATH/libcudart.so.12 not found in packages, sysroot(s) nor the missing_dso_whitelist.\n.. is this binary repackaging?']
```

So in the future we will be able to catch this kind of issue on CI

**Pull request merged**